File: packages/app/headless/backbone/rust/src/import_export.rs
use std::collections::HashMap;

use axum::{
    Json,
    extract::{Query, State},
    http::HeaderMap,
};
use serde::{Deserialize, Serialize};
use serde_json::Value;

use crate::db;
use crate::handlers::AppState;
use crate::models::*;

#[derive(Debug, Serialize, Deserialize)]
pub struct ExportData {
    pub collections: Vec<Collection>,
    pub records: HashMap<String, Vec<Record>>,
    pub buckets: Vec<Bucket>,
    pub files: Vec<FileRecord>,
}

#[derive(Debug, Deserialize)]
pub struct ImportQuery {
    pub skip_existing: Option<bool>,
}

pub async fn handle_export(
    headers: HeaderMap,
    State(state): State<std::sync::Arc<AppState>>,
) -> std::result::Result<Json<ExportData>, AppError> {
    crate::handlers::extract_claims(&headers)?;
    let conn = state.db.get().await.map_err(|e| AppError::Internal(e.to_string()))?;

    let collections = db::list_collections(&conn)?;
    let mut records = HashMap::new();
    for col in &collections {
        let page = db::list_records(&conn, &col.name, 1, 100000, "")?;
        records.insert(col.name.clone(), page.records);
    }
    let buckets = db::list_buckets(&conn)?;
    let mut files = Vec::new();
    for bucket in &buckets {
        let page = db::list_files(&conn, &bucket.name, 1, 100000)?;
        files.extend(page.files);
    }
    Ok(Json(ExportData {
        collections,
        records,
        buckets,
        files,
    }))
}

pub async fn handle_import(
    headers: HeaderMap,
    State(state): State<std::sync::Arc<AppState>>,
    Query(query): Query<ImportQuery>,
    Json(data): Json<ExportData>,
) -> std::result::Result<Json<Value>, AppError> {
    crate::handlers::extract_claims(&headers)?;
    let skip_existing = query.skip_existing.unwrap_or(false);
    let conn = state.db.get().await.map_err(|e| AppError::Internal(e.to_string()))?;

    let mut created_collections = 0i64;
    let mut created_records = 0i64;
    let mut created_buckets = 0i64;
    let mut created_files = 0i64;

    for col in &data.collections {
        let existing = db::get_collection(&conn, &col.name)?;
        if existing.is_some() {
            if skip_existing {
                continue;
            }
            return Err(AppError::Conflict(format!("collection '{}' already exists", col.name)));
        }
        db::insert_collection(&conn, &col.name, &col.schema)?;
        db::create_collection_table(&conn, &col.name, &col.schema)?;
        created_collections += 1;
    }

    for (col_name, recs) in &data.records {
        for rec in recs {
            let existing = db::get_record(&conn, col_name, &rec.id)?;
            if existing.is_some() {
                if skip_existing {
                    continue;
                }
                continue;
            }
            db::insert_record(&conn, col_name, &rec.id, &rec.data)?;
            created_records += 1;
        }
    }

    for bucket in &data.buckets {
        let existing = db::get_bucket(&conn, &bucket.name)?;
        if existing.is_some() {
            if skip_existing {
                continue;
            }
            return Err(AppError::Conflict(format!("bucket '{}' already exists", bucket.name)));
        }
        db::insert_bucket(&conn, &bucket.name, bucket.is_public)?;
        created_buckets += 1;
    }

    for f in &data.files {
        let existing = db::get_file(&conn, &f.id)?;
        if existing.is_some() {
            if skip_existing {
                continue;
            }
            continue;
        }
        db::insert_file(&conn, &f.bucket, &f.id, &f.filename, &f.mime_type, f.size)?;
        created_files += 1;
    }

    Ok(Json(serde_json::json!({
        "collections": created_collections,
        "records": created_records,
        "buckets": created_buckets,
        "files": created_files,
    })))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn export_data_empty_round_trip() {
        let data = ExportData {
            collections: vec![],
            records: HashMap::new(),
            buckets: vec![],
            files: vec![],
        };
        let json = serde_json::to_string(&data).unwrap();
        let deserialized: ExportData = serde_json::from_str(&json).unwrap();
        assert!(deserialized.collections.is_empty());
        assert!(deserialized.records.is_empty());
        assert!(deserialized.buckets.is_empty());
        assert!(deserialized.files.is_empty());
    }

    #[test]
    fn export_data_with_data_round_trip() {
        let data = ExportData {
            collections: vec![Collection {
                name: "users".into(),
                schema: "{\"name\":\"string\"}".into(),
                created_at: "2024-01-01T00:00:00Z".into(),
                updated_at: "2024-01-01T00:00:00Z".into(),
            }],
            records: {
                let mut map = HashMap::new();
                map.insert(
                    "users".into(),
                    vec![Record {
                        id: "rec_1".into(),
                        data: json!({"name": "Alice"}),
                        created_at: "2024-01-01T00:00:00Z".into(),
                        updated_at: "2024-01-01T00:00:00Z".into(),
                    }],
                );
                map
            },
            buckets: vec![Bucket {
                name: "files".into(),
                is_public: true,
                created_at: "2024-01-01T00:00:00Z".into(),
                updated_at: "2024-01-01T00:00:00Z".into(),
            }],
            files: vec![FileRecord {
                id: "f1".into(),
                bucket: "files".into(),
                filename: "test.txt".into(),
                mime_type: "text/plain".into(),
                size: 100,
                created_at: "2024-01-01T00:00:00Z".into(),
                updated_at: "2024-01-01T00:00:00Z".into(),
            }],
        };
        let json = serde_json::to_string(&data).unwrap();
        let deserialized: ExportData = serde_json::from_str(&json).unwrap();
        assert_eq!(deserialized.collections.len(), 1);
        assert_eq!(deserialized.collections[0].name, "users");
        assert_eq!(deserialized.records.len(), 1);
        assert_eq!(deserialized.records["users"].len(), 1);
        assert_eq!(deserialized.records["users"][0].data["name"], "Alice");
        assert_eq!(deserialized.buckets.len(), 1);
        assert_eq!(deserialized.buckets[0].name, "files");
        assert!(deserialized.buckets[0].is_public);
        assert_eq!(deserialized.files.len(), 1);
        assert_eq!(deserialized.files[0].filename, "test.txt");
        assert_eq!(deserialized.files[0].size, 100);
    }

    #[test]
    fn export_data_json_structure() {
        let data = ExportData {
            collections: vec![],
            records: HashMap::new(),
            buckets: vec![],
            files: vec![],
        };
        let value = serde_json::to_value(&data).unwrap();
        assert!(value.is_object());
        assert!(value.get("collections").unwrap().is_array());
        assert!(value.get("records").unwrap().is_object());
        assert!(value.get("buckets").unwrap().is_array());
        assert!(value.get("files").unwrap().is_array());
        assert_eq!(value["collections"].as_array().unwrap().len(), 0);
        assert_eq!(value["records"].as_object().unwrap().len(), 0);
        assert_eq!(value["buckets"].as_array().unwrap().len(), 0);
        assert_eq!(value["files"].as_array().unwrap().len(), 0);
    }

    #[test]
    fn import_query_default_skip_existing() {
        let query: ImportQuery = serde_json::from_str("{}").unwrap();
        assert_eq!(query.skip_existing, None);
    }

    #[test]
    fn import_query_skip_existing_true() {
        let query: ImportQuery = serde_json::from_str(r#"{"skip_existing": true}"#).unwrap();
        assert_eq!(query.skip_existing, Some(true));
    }

    #[test]
    fn import_query_skip_existing_false() {
        let query: ImportQuery = serde_json::from_str(r#"{"skip_existing": false}"#).unwrap();
        assert_eq!(query.skip_existing, Some(false));
    }
}

File: packages/app/headless/backbone/languages/go/internal/httpapi/importexport.go
package httpapi

import (
	"encoding/json"
	"fmt"
	"net/http"

	"github.com/hieudoanm/backbone/internal/id"
	"github.com/hieudoanm/backbone/internal/log"
	"github.com/hieudoanm/backbone/internal/store"
)

func (s *Server) handleExport(w http.ResponseWriter, r *http.Request) {
	cols, err := store.ListCollections(s.db)
	if err != nil {
		errorJSON(w, err.Error(), http.StatusInternalServerError)
		return
	}

	records := make(map[string][]store.Record)
	for _, c := range cols {
		page, err := store.ListRecords(s.db, c.Name, 1, 1000000, nil, "", nil, "")
		if err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
		records[c.Name] = page.Records
	}

	buckets, err := store.ListBuckets(s.db)
	if err != nil {
		errorJSON(w, err.Error(), http.StatusInternalServerError)
		return
	}

	var files []store.FileRecord
	for _, b := range buckets {
		page, err := store.ListFiles(s.db, b.Name, 1, 1000000)
		if err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
		files = append(files, page.Files...)
	}

	jsonResponse(w, map[string]any{
		"collections": cols,
		"records":     records,
		"buckets":     buckets,
		"files":       files,
	})
}

func (s *Server) handleImport(w http.ResponseWriter, r *http.Request) {
	skipExisting := r.URL.Query().Get("skip_existing") == "true"

	var body struct {
		Collections []store.Collection        `json:"collections"`
		Records     map[string][]store.Record `json:"records"`
		Buckets     []store.Bucket            `json:"buckets"`
		Files       []store.FileRecord        `json:"files"`
	}
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		errorJSON(w, "invalid json", http.StatusBadRequest)
		return
	}

	for _, c := range body.Collections {
		existing, err := store.GetCollection(s.db, c.Name)
		if err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
		if existing != nil {
			if !skipExisting {
				errorJSON(w, fmt.Sprintf("collection %q already exists", c.Name), http.StatusConflict)
				return
			}
			continue
		}
		if err := store.CreateCollection(s.db, c.Name, c.Schema); err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
	}

	for colName, recs := range body.Records {
		for _, rec := range recs {
			existing, err := store.GetRecord(s.db, colName, rec.ID)
			if err != nil {
				errorJSON(w, err.Error(), http.StatusInternalServerError)
				return
			}
			if existing != nil {
				if !skipExisting {
					errorJSON(w, fmt.Sprintf("record %q in collection %q already exists", rec.ID, colName), http.StatusConflict)
					return
				}
				continue
			}
			if _, err := store.CreateRecord(s.db, colName, rec.ID, rec.Data); err != nil {
				errorJSON(w, err.Error(), http.StatusInternalServerError)
				return
			}
		}
	}

	for _, b := range body.Buckets {
		existing, err := store.GetBucket(s.db, b.Name)
		if err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
		if existing != nil {
			if !skipExisting {
				errorJSON(w, fmt.Sprintf("bucket %q already exists", b.Name), http.StatusConflict)
				return
			}
			continue
		}
		if _, err := store.CreateBucket(s.db, b.Name, b.IsPublic); err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
	}

	for _, f := range body.Files {
		existing, err := store.GetFile(s.db, f.ID)
		if err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
		if existing != nil {
			if !skipExisting {
				errorJSON(w, fmt.Sprintf("file %q already exists", f.ID), http.StatusConflict)
				return
			}
			continue
		}
		if _, err := store.InsertFile(s.db, f.Bucket, f.ID, f.Filename, f.MimeType, f.Size); err != nil {
			errorJSON(w, err.Error(), http.StatusInternalServerError)
			return
		}
	}

	meta, _ := json.Marshal(map[string]any{
		"import":        true,
		"collections":   len(body.Collections),
		"records":       len(body.Records),
		"buckets":       len(body.Buckets),
		"files":         len(body.Files),
		"skip_existing": skipExisting,
	})
	if _, err := log.InsertLog(s.db, id.Generate(), "info", "import completed", string(meta)); err != nil {
		errorJSON(w, err.Error(), http.StatusInternalServerError)
		return
	}

	jsonResponse(w, map[string]string{"status": "imported"})
}
